File: bindings/typescript/src/product.ts
import type { Signer } from "./workflow.js";
import type { AttachedAgent, Profile } from "./workflow.js";
import {
  executeMcpClosed,
  executeMcpPlanClosed,
  recoverMcpClosed,
  resumeMcpClosed,
  resourcesForMcpAuthority,
  type McpAction,
  type McpClosedProvider,
  type McpExecutionState,
  type McpExecutionObserver,
  type McpReceiptSink,
  type McpAttestedReceipt,
  type McpToolAuthority,
  type McpPlanClosedResult,
} from "./profiles/mcp/index.js";
import type { ProfilePlan } from "./plans.js";
import type { ApplicationReceiptAttestor } from "./profiles/application/index.js";
import {
  decodeLinkedReceipt,
  encodeLinkedReceipt,
  verifyLinkedReceipt,
} from "./internal/receipt-attestation.js";
import {
  createProductionAuths,
  type ProductionAuths,
  type ProductionAuthsOptions,
} from "./production-client.js";

const configurationResources = new WeakMap<AuthsConfiguration, InternalConfiguration>();
const referenceResources = new WeakMap<ExecutionReference, string>();
type RawMcpExecution = Awaited<ReturnType<typeof executeMcpClosed>>;
type RawMcpPlanExecution = Awaited<ReturnType<typeof executeMcpPlanClosed>>;

export interface Actor {
  readonly principal: string;
}

export type Authority = McpToolAuthority;

export interface AuthsConfiguration {
  readonly mode: "development" | "production";
  readonly diagnostics: readonly string[];
}

export type Receipt = McpAttestedReceipt;

export interface Completed {
  readonly kind: "completed";
  readonly executionId: string;
  readonly result: unknown;
  readonly receipt: Receipt;
}

export interface PlanCompleted {
  readonly kind: "completed";
  readonly results: readonly unknown[];
  readonly receipts: readonly Receipt[];
}

export interface PlanRecoveryResult {
  readonly kind: "recoverable" | "not-applied" | "exact-replay" | "conflict";
  readonly executionId: string;
  readonly completedResults: readonly unknown[];
  readonly completedReceipts: readonly Receipt[];
  readonly reference?: ExecutionReference;
}

export interface Denied {
  readonly kind: "denied";
  readonly code: string;
}

export interface Indeterminate {
  readonly kind: "indeterminate";
  readonly code: string;
}

export class ExecutionReference {
  private constructor(token: symbol, value: string) {
    if (token !== REFERENCE_TOKEN) throw new TypeError("sealed Auths execution reference");
    referenceResources.set(this, value);
    Object.freeze(this);
  }

  static create(token: symbol, value: string): ExecutionReference {
    return new ExecutionReference(token, value);
  }

  static decode(input: Uint8Array): ExecutionReference {
    return decodeExecutionReference(input);
  }

  encode(): Uint8Array {
    return encodeExecutionReference(this);
  }

  toJSON(): never {
    throw new TypeError("Auths execution references are not serializable");
  }
}

export function encodeExecutionReference(reference: ExecutionReference): Uint8Array {
  const value = referenceResources.get(reference);
  if (value === undefined) throw new TypeError("forged Auths execution reference");
  return new TextEncoder().encode(value);
}

export function decodeExecutionReference(input: Uint8Array): ExecutionReference {
  if (!(input instanceof Uint8Array) || input.length !== 134) {
    throw new TypeError("invalid Auths execution reference");
  }
  const value = new TextDecoder("utf-8", { fatal: true }).decode(input);
  if (!/^mcp1\.[0-9a-f]{64}\.[0-9a-f]{64}$/.test(value)) {
    throw new TypeError("invalid Auths execution reference");
  }
  return ExecutionReference.create(REFERENCE_TOKEN, value);
}

const REFERENCE_TOKEN = Symbol("auths-execution-reference");

export interface RecoveryResult {
  readonly kind: "recoverable" | "not-applied" | "exact-replay" | "conflict";
  readonly executionId: string;
  readonly reference?: ExecutionReference;
}

export type ExecutionResult = Completed | PlanCompleted | Denied | Indeterminate | RecoveryResult | PlanRecoveryResult;
export type SingleExecutionResult = Completed | Denied | Indeterminate | RecoveryResult;
export type PlanExecutionResult = PlanCompleted | Denied | Indeterminate | PlanRecoveryResult;

export interface Auths {
  readonly actor: Actor;
  readonly authority: Authority;
  readonly diagnostics: readonly string[];
  execute(input: Readonly<{
    action: McpAction;
    provider: McpClosedProvider;
    requestId?: string;
  }>): Promise<SingleExecutionResult>;
  execute(input: Readonly<{
    plan: ProfilePlan<McpAction>;
    provider: McpClosedProvider;
    requestId?: string;
  }>): Promise<PlanExecutionResult>;
  resume(input: Readonly<{
    reference: ExecutionReference;
    provider: McpClosedProvider;
  }>): Promise<SingleExecutionResult>;
  recover(input: Readonly<{
    action: McpAction;
    provider: McpClosedProvider;
    requestId?: string;
  }>): Promise<SingleExecutionResult>;
  delegate(input: Readonly<{
    authority: McpToolAuthority;
    name?: string;
    expiresInSeconds?: number;
  }>): Promise<Auths>;
  close(): Promise<void>;
  [Symbol.asyncDispose](): Promise<void>;
}

interface InternalConfiguration extends AuthsConfiguration {
  open(): Promise<AuthsResources>;
}

export interface AuthsResources {
  readonly agent: AttachedAgent<Profile>;
  readonly authority: McpToolAuthority;
  readonly state: McpExecutionState;
  readonly receipts: McpReceiptSink;
  readonly receiptAttestor: ApplicationReceiptAttestor;
  readonly sessionKey: Uint8Array;
  readonly childSigner: () => Promise<Signer>;
  readonly observer?: McpExecutionObserver;
  readonly dispose: () => Promise<void>;
}

class AuthsFacade implements Auths {
  readonly actor: Actor;
  readonly authority: Authority;
  readonly diagnostics: readonly string[];
  readonly #resources: AuthsResources;
  readonly #children = new Set<AuthsFacade>();
  #closed = false;

  constructor(resources: AuthsResources, diagnostics: readonly string[]) {
    this.#resources = resources;
    this.actor = Object.freeze({ principal: resources.agent.identity.principal.principal });
    this.authority = resources.authority;
    this.diagnostics = Object.freeze([...diagnostics]);
  }

  async execute(input: Readonly<{
    action: McpAction;
    provider: McpClosedProvider;
    requestId?: string;
  }>): Promise<SingleExecutionResult>;
  async execute(input: Readonly<{
    plan: ProfilePlan<McpAction>;
    provider: McpClosedProvider;
    requestId?: string;
  }>): Promise<PlanExecutionResult>;
  async execute(input: Readonly<{
    action?: McpAction;
    plan?: ProfilePlan<McpAction>;
    provider: McpClosedProvider;
    requestId?: string;
  }>): Promise<ExecutionResult> {
    this.#assertActive();
    this.#assertProvider(input.provider);
    const execution = {
      provider: input.provider,
      state: this.#resources.state,
      receipts: this.#resources.receipts,
      attestor: this.#resources.receiptAttestor,
      sessionKey: this.#resources.sessionKey,
      ...(this.#resources.observer === undefined ? {} : { observer: this.#resources.observer }),
      ...(input.requestId === undefined ? {} : { requestId: input.requestId }),
    };
    if (input.action !== undefined && input.plan === undefined) {
      return projectExecution(await executeMcpClosed(this.#resources.agent, input.action, execution));
    }
    if (input.plan !== undefined && input.action === undefined) {
      return projectPlanExecution(await executeMcpPlanClosed(this.#resources.agent, input.plan, execution));
    }
    throw new TypeError("Auths execute requires exactly one action or plan");
  }

  async resume(input: Readonly<{
    reference: ExecutionReference;
    provider: McpClosedProvider;
  }>): Promise<SingleExecutionResult> {
    this.#assertActive();
    this.#assertProvider(input.provider);
    const reference = referenceResources.get(input.reference);
    if (reference === undefined) throw new TypeError("forged Auths execution reference");
    return projectExecution(await resumeMcpClosed(
      this.#resources.agent,
      reference,
      {
        provider: input.provider,
        state: this.#resources.state,
        receipts: this.#resources.receipts,
        attestor: this.#resources.receiptAttestor,
        sessionKey: this.#resources.sessionKey,
        ...(this.#resources.observer === undefined ? {} : { observer: this.#resources.observer }),
      },
    ));
  }

  async recover(input: Readonly<{
    action: McpAction;
    provider: McpClosedProvider;
    requestId?: string;
  }>): Promise<SingleExecutionResult> {
    this.#assertActive();
    this.#assertProvider(input.provider);
    return projectExecution(await recoverMcpClosed(
      this.#resources.agent,
      input.action,
      {
        provider: input.provider,
        state: this.#resources.state,
        receipts: this.#resources.receipts,
        attestor: this.#resources.receiptAttestor,
        sessionKey: this.#resources.sessionKey,
        ...(this.#resources.observer === undefined ? {} : { observer: this.#resources.observer }),
        ...(input.requestId === undefined ? {} : { requestId: input.requestId }),
      },
    ));
  }

  async delegate(input: Readonly<{
    authority: McpToolAuthority;
    name?: string;
    expiresInSeconds?: number;
  }>): Promise<Auths> {
    this.#assertActive();
    const authority = resourcesForMcpAuthority(input.authority);
    const current = resourcesForMcpAuthority(this.authority);
    if (authority.profile.service !== current.profile.service) {
      throw new TypeError("delegated authority belongs to another MCP service");
    }
    const expiresInSeconds = input.expiresInSeconds ?? 300;
    if (!Number.isSafeInteger(expiresInSeconds) || expiresInSeconds < 1 || expiresInSeconds > 86_400) {
      throw new TypeError("delegated authority expiry is outside bounds");
    }
    const now = BigInt(Math.floor(Date.now() / 1000));
    const signer = await this.#resources.childSigner();
    const agent = await this.#resources.agent.delegate({
      name: input.name ?? "delegated-agent",
      signer,
      authority: {
        permissions: authority.permissions,
        validity: { notBefore: now, expiresAt: now + BigInt(expiresInSeconds) },
        audiences: authority.audiences,
        remainingDepth: 0,
        actionConstraint: { kind: "inherit" },
        budget: { kind: "inherit" },
        status: { kind: "expiry-only" },
      },
    });
    const child = new AuthsFacade({
      ...this.#resources,
      agent,
      authority: input.authority,
      dispose: () => agent.dispose(),
    }, this.diagnostics);
    this.#children.add(child);
    return child;
  }

  async close(): Promise<void> {
    if (this.#closed) return;
    this.#closed = true;
    for (const child of this.#children) await child.close();
    this.#children.clear();
    await this.#resources.dispose();
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close();
  }

  #assertActive(): void {
    if (this.#closed) throw new TypeError("Auths is closed");
  }

  #assertProvider(provider: McpClosedProvider): void {
    const authority = resourcesForMcpAuthority(this.authority);
    if (provider?.profile !== "auths.mcp" || provider.service !== authority.profile.service) {
      throw new TypeError("MCP provider does not match this Auths authority");
    }
  }
}

export function createAuthsConfiguration(
  mode: "development" | "production",
  diagnostics: readonly string[],
  open: () => Promise<AuthsResources>,
): AuthsConfiguration {
  if (typeof open !== "function" || diagnostics.some((value) => typeof value !== "string" || value.length === 0 || value.length > 256)) {
    throw new TypeError("invalid Auths composition");
  }
  const configuration = Object.freeze({ mode, diagnostics: Object.freeze([...diagnostics]) });
  configurationResources.set(configuration, { ...configuration, open });
  return configuration;
}

export function createAuths(configuration: AuthsConfiguration): Promise<Auths>;
export function createAuths(configuration: ProductionAuthsOptions): ProductionAuths;
export function createAuths(
  configuration: AuthsConfiguration | ProductionAuthsOptions,
): Promise<Auths> | ProductionAuths {
  if ("endpoint" in configuration) return createProductionAuths(configuration);
  const resources = configurationResources.get(configuration);
  if (resources === undefined) throw new TypeError("Auths configuration was not created by an integration");
  return resources.open().then((opened) => new AuthsFacade(opened, resources.diagnostics));
}

export async function verifyReceipt(receipt: Receipt): Promise<void> {
  await verifyLinkedReceipt(receipt);
}

export function encodeReceipt(receipt: Receipt): Uint8Array {
  return encodeLinkedReceipt(receipt);
}

export function decodeReceipt(input: Uint8Array): Receipt {
  return decodeLinkedReceipt(input);
}

function projectExecution(value: RawMcpExecution): SingleExecutionResult {
  if (value.kind === "completed") {
    return Object.freeze({
      kind: "completed" as const,
      executionId: value.executionId,
      result: value.result,
      receipt: value.receipt,
    });
  }
  if (value.kind === "denied" || value.kind === "indeterminate") {
    return Object.freeze({ kind: value.kind, code: value.code });
  }
  if (value.kind === "recoverable") {
    return Object.freeze({
      kind: "recoverable" as const,
      executionId: value.executionId,
      reference: ExecutionReference.create(REFERENCE_TOKEN, value.executionReference),
    });
  }
  return Object.freeze({ kind: value.kind, executionId: value.executionId });
}

function projectPlanExecution(value: RawMcpPlanExecution): PlanExecutionResult {
  if ("failedIndex" in value) {
    return Object.freeze({ kind: value.kind, code: value.result.code });
  }
  if (value.kind === "completed") {
    return Object.freeze({ kind: "completed", results: value.results, receipts: value.receipts });
  }
  return Object.freeze({
    kind: value.kind,
    executionId: value.executionId,
    completedResults: value.completedResults,
    completedReceipts: value.completedReceipts,
    ...(value.kind === "recoverable"
      ? { reference: ExecutionReference.create(REFERENCE_TOKEN, value.executionReference) }
      : {}),
  });
}

File: bindings/python/python/auths/_product.py
from __future__ import annotations

import time
from dataclasses import dataclass
from types import TracebackType
from typing import Awaitable, Callable, Final, Literal, NoReturn, Optional, Type, Union

from ._product_errors import (
    AuthsError,
    AuthsErrorCode,
    EffectState,
    RecommendedAction,
    RetryClass,
    classify,
)
from .profiles._mcp import (
    McpAction,
    McpClosedProvider,
    McpCompleted,
    McpDenied,
    McpExecutionResources,
    McpExecutionObserver,
    McpExecutionStore,
    McpIndeterminate,
    McpNotApplied,
    McpPlan,
    McpPlanCompleted,
    McpPlanDenied,
    McpPlanIndeterminate,
    McpPlanRecoveryResult,
    McpReceiptSink,
    McpRecoverable,
    McpToolAuthority,
    execute_mcp_closed,
    execute_mcp_plan_closed,
    resources_for_mcp_authority,
    resume_mcp_closed,
)
from ._receipts import (
    Receipt,
    ReceiptAttestor,
    decode_linked_receipt,
    encode_linked_receipt,
    verify_linked_receipt,
)
from ._workflow import (
    AttachedAgent,
    DelegatedAuthority,
    ExpiryOnly,
    InheritAction,
    InheritBudget,
    Signer,
    Validity,
)

_CONFIGURATION_TOKEN = object()
_REFERENCE_TOKEN = object()


@dataclass(frozen=True)
class Actor:
    principal: str


Authority = McpToolAuthority


@dataclass(frozen=True)
class Completed:
    kind: Literal["completed"]
    execution_id: str
    result: object
    receipt: Receipt


@dataclass(frozen=True)
class Denied:
    """Nothing happened, and the caller can prove it from `effect`."""

    kind: Literal["denied"]
    code: AuthsErrorCode
    reason: str
    """The kernel's own denial reason, e.g. `permission-not-granted`.

    Diagnostic. `code` is the stable identity a caller branches on.
    """
    effect: EffectState
    retry: RetryClass
    recommended_action: RecommendedAction


@dataclass(frozen=True)
class Indeterminate:
    """The decision could not be reached; read `effect` before retrying."""

    kind: Literal["indeterminate"]
    code: AuthsErrorCode
    reason: str
    effect: EffectState
    retry: RetryClass
    recommended_action: RecommendedAction


class ExecutionReference:
    __slots__ = ("_value",)

    def __init__(self, token: object, value: str) -> None:
        if token is not _REFERENCE_TOKEN:
            raise TypeError("sealed Auths execution reference")
        self._value = value

    def __copy__(self) -> None:
        raise TypeError("Auths execution reference is not copyable")

    def __deepcopy__(self, _: object) -> None:
        raise TypeError("Auths execution reference is not copyable")

    def __reduce__(self) -> NoReturn:
        raise TypeError("Auths execution reference is not serializable")

    @classmethod
    def from_bytes(cls, value: bytes) -> ExecutionReference:
        return decode_execution_reference(value)

    def to_bytes(self) -> bytes:
        return encode_execution_reference(self)


def encode_execution_reference(reference: ExecutionReference) -> bytes:
    if type(reference) is not ExecutionReference:
        raise TypeError("forged Auths execution reference")
    return reference._value.encode()


def decode_execution_reference(value: bytes) -> ExecutionReference:
    try:
        parsed = bytes(value).decode("ascii")
    except (TypeError, UnicodeDecodeError):
        raise ValueError("invalid Auths execution reference") from None
    if (
        len(parsed) != 134
        or not parsed.startswith("mcp1.")
        or parsed[69] != "."
        or any(character not in "0123456789abcdef" for character in parsed[5:69])
        or any(character not in "0123456789abcdef" for character in parsed[70:])
    ):
        raise ValueError("invalid Auths execution reference")
    return ExecutionReference(_REFERENCE_TOKEN, parsed)


@dataclass(frozen=True)
class RecoveryResult:
    """An execution that did not complete.

    `effect` is the safety-critical field: `possible` means the real-world
    effect may already have been applied and a blind retry may repeat it.
    """

    kind: Literal["recoverable", "not-applied", "exact-replay", "conflict"]
    execution_id: str
    code: AuthsErrorCode
    effect: EffectState
    retry: RetryClass
    recommended_action: RecommendedAction
    reference: Optional[ExecutionReference] = None


@dataclass(frozen=True)
class PlanCompleted:
    kind: Literal["completed"]
    results: tuple[object, ...]
    receipts: tuple[Receipt, ...]


@dataclass(frozen=True)
class PlanRecoveryResult:
    kind: Literal["recoverable", "not-applied", "exact-replay", "conflict"]
    execution_id: str
    code: AuthsErrorCode
    effect: EffectState
    retry: RetryClass
    recommended_action: RecommendedAction
    completed_results: tuple[object, ...]
    completed_receipts: tuple[Receipt, ...]
    reference: Optional[ExecutionReference] = None


ExecutionResult = Union[
    Completed,
    PlanCompleted,
    Denied,
    Indeterminate,
    RecoveryResult,
    PlanRecoveryResult,
]


@dataclass(frozen=True)
class _AuthsResources:
    agent: AttachedAgent
    authority: McpToolAuthority
    state: McpExecutionStore
    receipts: McpReceiptSink
    receipt_attestor: ReceiptAttestor
    session_key: bytes
    child_signer: Callable[[], Awaitable[Signer]]
    dispose: Callable[[], Awaitable[None]]
    observer: Optional[McpExecutionObserver] = None


class AuthsConfiguration:
    __slots__ = ("mode", "diagnostics", "_open")

    def __init__(
        self,
        token: object,
        mode: Literal["development", "production"],
        diagnostics: tuple[str, ...],
        open_resources: Callable[[], Awaitable[_AuthsResources]],
    ) -> None:
        if token is not _CONFIGURATION_TOKEN:
            raise TypeError("sealed Auths configuration")
        self.mode = mode
        self.diagnostics = diagnostics
        self._open = open_resources


class Auths:
    def __init__(
        self,
        resources: _AuthsResources,
        diagnostics: tuple[str, ...],
    ) -> None:
        self._resources = resources
        self.actor = Actor(resources.agent.identity.principal.principal.value)
        self.authority: Authority = resources.authority
        self.diagnostics = diagnostics
        self._children: set[Auths] = set()
        self._closed = False

    async def execute(
        self,
        *,
        action: Optional[McpAction] = None,
        plan: Optional[McpPlan] = None,
        provider: McpClosedProvider,
        request_id: Optional[str] = None,
    ) -> ExecutionResult:
        self._assert_active()
        self._assert_provider(provider)
        execution = McpExecutionResources(
            provider,
            self._resources.state,
            self._resources.receipts,
            self._resources.receipt_attestor,
            self._resources.session_key,
            request_id,
            self._resources.observer,
        )
        if action is not None and plan is None:
            return _project_execution(
                await execute_mcp_closed(self._resources.agent, action, execution)
            )
        if plan is not None and action is None:
            return _project_plan_execution(
                await execute_mcp_plan_closed(self._resources.agent, plan, execution)
            )
        raise TypeError("Auths execute requires exactly one action or plan")

    async def resume(
        self,
        *,
        reference: ExecutionReference,
        provider: McpClosedProvider,
    ) -> ExecutionResult:
        self._assert_active()
        self._assert_provider(provider)
        if type(reference) is not ExecutionReference:
            raise TypeError("forged Auths execution reference")
        result = await resume_mcp_closed(
            reference._value,
            McpExecutionResources(
                provider,
                self._resources.state,
                self._resources.receipts,
                self._resources.receipt_attestor,
                self._resources.session_key,
                None,
                self._resources.observer,
            ),
        )
        return _project_execution(result)

    async def delegate(
        self,
        *,
        authority: McpToolAuthority,
        name: str = "delegated-agent",
        expires_in_seconds: int = 300,
    ) -> Auths:
        self._assert_active()
        profile, permissions, _, audiences = resources_for_mcp_authority(authority)
        parent_profile, _, _, _ = resources_for_mcp_authority(self.authority)
        if profile.service != parent_profile.service:
            raise TypeError("delegated authority belongs to another MCP service")
        if (
            type(expires_in_seconds) is not int
            or expires_in_seconds < 1
            or expires_in_seconds > 86_400
        ):
            raise ValueError("delegated authority expiry is outside bounds")
        now = int(time.time())
        signer = await self._resources.child_signer()
        agent = await self._resources.agent.delegate(
            name=name,
            authority=DelegatedAuthority(
                permissions,
                Validity(now, now + expires_in_seconds),
                audiences,
                0,
                InheritAction(),
                InheritBudget(),
                ExpiryOnly(),
            ),
            signer=signer,
        )

        async def dispose() -> None:
            await agent.aclose()

        child = Auths(
            _AuthsResources(
                agent,
                authority,
                self._resources.state,
                self._resources.receipts,
                self._resources.receipt_attestor,
                self._resources.session_key,
                self._resources.child_signer,
                dispose,
                self._resources.observer,
            ),
            self.diagnostics,
        )
        self._children.add(child)
        return child

    async def aclose(self) -> None:
        if self._closed:
            return
        self._closed = True
        for child in tuple(self._children):
            await child.aclose()
        self._children.clear()
        await self._resources.dispose()

    async def __aenter__(self) -> Auths:
        self._assert_active()
        return self

    async def __aexit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc: Optional[BaseException],
        traceback: Optional[TracebackType],
    ) -> None:
        await self.aclose()

    def _assert_active(self) -> None:
        if self._closed:
            raise RuntimeError("Auths is closed")

    def _assert_provider(self, provider: McpClosedProvider) -> None:
        profile, _, _, _ = resources_for_mcp_authority(self.authority)
        if (
            getattr(provider, "profile", None) != "auths.mcp"
            or getattr(provider, "service", None) != profile.service
        ):
            raise TypeError("MCP provider does not match this Auths authority")


def _create_auths_configuration(
    mode: Literal["development", "production"],
    diagnostics: tuple[str, ...],
    open_resources: Callable[[], Awaitable[_AuthsResources]],
) -> AuthsConfiguration:
    if (
        not diagnostics
        or any(
            type(value) is not str or not value or len(value) > 256
            for value in diagnostics
        )
        or not callable(open_resources)
    ):
        raise TypeError("invalid Auths composition")
    return AuthsConfiguration(
        _CONFIGURATION_TOKEN,
        mode,
        tuple(diagnostics),
        open_resources,
    )


async def _create_auths(configuration: AuthsConfiguration) -> Auths:
    if type(configuration) is not AuthsConfiguration:
        raise TypeError("Auths configuration was not created by an integration")
    return Auths(await configuration._open(), configuration.diagnostics)


def verify_receipt(receipt: Receipt) -> None:
    verify_linked_receipt(receipt)


def encode_receipt(receipt: Receipt) -> bytes:
    return encode_linked_receipt(receipt)


def decode_receipt(value: bytes) -> Receipt:
    return decode_linked_receipt(value)


# The two registry codes that name a kernel verdict. The kernel's own reason
# string (`permission-not-granted`, ...) is carried as `reason`; these are the
# stable identities whose effect, retry class, and recommended action Rust owns.
_DENIED_CODE: Final = "core.authorization-denied"
_INDETERMINATE_CODE: Final = "core.authorization-indeterminate"


def _axis(code: str) -> tuple[EffectState, RetryClass, RecommendedAction]:
    """Reads Rust's classification of `code`. Nothing here is computed."""
    classification = classify(code)
    return (
        classification.effect,
        classification.retry,
        classification.recommended_action,
    )


def _project_execution(value: object) -> ExecutionResult:
    if isinstance(value, McpCompleted):
        return Completed(
            "completed",
            value.execution_id,
            value.result,
            Receipt(value.receipt.decision, value.receipt.execution),
        )
    if isinstance(value, McpDenied):
        return Denied("denied", _DENIED_CODE, value.code, *_axis(_DENIED_CODE))
    if isinstance(value, McpIndeterminate):
        return Indeterminate(
            "indeterminate",
            _INDETERMINATE_CODE,
            value.code,
            *_axis(_INDETERMINATE_CODE),
        )
    if isinstance(value, McpRecoverable):
        return RecoveryResult(
            "recoverable",
            value.execution_id,
            value.code,
            *_axis(value.code),
            ExecutionReference(_REFERENCE_TOKEN, value.execution_reference),
        )
    if isinstance(value, McpNotApplied):
        return RecoveryResult(
            value.kind, value.execution_id, value.code, *_axis(value.code)
        )
    raise RuntimeError("MCP execution returned an unsupported result")


def _project_plan_execution(value: object) -> ExecutionResult:
    if isinstance(value, McpPlanCompleted):
        return PlanCompleted(
            "completed",
            value.results,
            tuple(Receipt(item.decision, item.execution) for item in value.receipts),
        )
    if isinstance(value, McpPlanDenied):
        return Denied("denied", _DENIED_CODE, value.result.code, *_axis(_DENIED_CODE))
    if isinstance(value, McpPlanIndeterminate):
        return Indeterminate(
            "indeterminate",
            _INDETERMINATE_CODE,
            value.result.code,
            *_axis(_INDETERMINATE_CODE),
        )
    if isinstance(value, McpPlanRecoveryResult):
        return PlanRecoveryResult(
            value.kind,
            value.execution_id,
            value.code,
            *_axis(value.code),
            value.completed_results,
            tuple(
                Receipt(item.decision, item.execution)
                for item in value.completed_receipts
            ),
            ExecutionReference(_REFERENCE_TOKEN, value.execution_reference)
            if value.execution_reference is not None
            else None,
        )
    raise RuntimeError("MCP plan execution returned an unsupported result")


__all__ = [
    "Actor",
    "Auths",
    "AuthsConfiguration",
    "AuthsError",
    "AuthsErrorCode",
    "Authority",
    "Completed",
    "Denied",
    "decode_execution_reference",
    "decode_receipt",
    "encode_execution_reference",
    "encode_receipt",
    "EffectState",
    "ExecutionReference",
    "ExecutionResult",
    "Indeterminate",
    "PlanCompleted",
    "PlanRecoveryResult",
    "Receipt",
    "RecommendedAction",
    "RecoveryResult",
    "RetryClass",
    "verify_receipt",
]
